Support remote action cache for nested local execution

juergbi opened the pull request:
This generalizes the `remote-execution` config section to support using the Remote Execution API while still using local execution, only using the CAS and Action Cache services of the REAPI. I.e., this allows a `remote-execution` config section with `action-cache-service` also if no `execution-service` is configured.

This branch supports a remote action cache with local execution only for nested REAPI clients using the remote execution protocol. Using an action cache (local or remote) also for BuildStream build jobs is not implemented in this branch but is planned for a follow-up.

This requires buildbox-casd master or 1.3.23 (should be released early next week).

juergbi committed:
Require buildbox-casd 1.2.6

This makes it possible to drop the fallback to direct connections to
remote servers.

juergbi committed:
sandbox: Create single buildbox-casd instance for remote execution

juergbi committed:
Support remote action cache for nested execution

juergbi committed:
tests/integration/sandbox.py: Add test for remote action cache config

juergbi committed:
Update protobuf to 5.29.5 and grpcio to 1.69.0

This is the latest protobuf version before protobuf 6, which includes
breaking changes. protobuf 5.29 is supported until 31 Mar 2026.

abderrahim commented on src/buildstream/_remote.py:
```diff
@@ -72,10 +71,6 @@ def init(self):
-            self.channel.close()
-            self.channel = None
-
         self._initialized = False
```
Is there any value in keeping the `close()` method now? AFAICT, there is no more cleanup to do now that everything goes through buildbox-casd.

juergbi replied:
I think it's indeed no longer necessary. Removed.

abderrahim commented on requirements/requirements.in:
```diff
@@ -4,7 +4,7 @@ Jinja2 >= 2.10
 packaging
 pluginbase
-protobuf <6.0dev,>=5.26.1
+protobuf <6.0dev,>=5.29
```
Personal preference, but I like to have updating .proto files and updating the grpcio-tools version that generates them in different commits.

juergbi replied:
Makes sense. Split into two commits.

abderrahim commented on src/buildstream/sandbox/_sandboxbuildboxrun.py:
```diff
@@ -32,6 +34,27 @@
+        cascache = context.get_cascache()
+
+        re_specs = context.remote_execution_specs
+        if re_specs and re_specs.action_spec and not re_specs.exec_spec:
```
Is there a value in excluding this when we have `re_specs.exec_spec`? IIUC, we can use this sandbox even though remote-execution is configured in code paths that go through `Element._prepare_sandbox()` (which sets allow_remote=False) such as `bst shell`.

Is there a reason to prevent using the action cache in that situation?

juergbi replied:
When we add action cache support for BuildStream actions, we need to make sure that we're not attempting to cache interactive shells. However, I don't think we need to disable remote execution or action cache access for sandboxed REAPI clients, so I've removed this condition.

abderrahim commented on src/buildstream/sandbox/_sandboxbuildboxrun.py:
```diff
@@ -91,8 +114,10 @@ def _execute_action(self, action, flags):
-            raise SandboxError("'remote-apis-socket' is not currently supported with 'storage-service'.")
+        if config.remote_apis_socket_path and context.remote_cache_spec and not self.re_remote:
+            raise SandboxError(
+                "'remote-apis-socket' is not supported with 'storage-service' without 'action-cache-service'."
```
So IIUC, we now support

* fully local: the buildbox-casd started by buildstream handles everything (storage, execution, action cache)
* fully remote: the remote execution servers handle everything (cache storage-service could double as storage for remote execution)
* local with remote action cache: execution is done locally, storage can be either the `cache` storage-service or the `remote-execution` storage-service, action cache has to be configured in `remote-execution`.

Assuming the above is correct, I find that this error message a bit confusing as you can't set the `action-cache-service` in the `cache` configuration.

juergbi replied:
I've adjusted the error message, hope it's clearer now.

abderrahim commented on tests/integration/sandbox.py:
```diff
@@ -59,3 +60,24 @@ def test_remote_apis_socket(cli, datafiles):
+            {
+                "remote-execution": {
+                    "storage-service": {"url": share.repo},
+                    "action-cache-service": {"url": share.repo, "push": True},
```
Could you please add tests for the other possible configurations? (cache storage-service with remote-execution action-cache-service, and both storage services with remote-execution action-cache-service)

Slightly related, considering the case where both the `cache` and `remote-execution` storage services are configured, does this lead to a sensible outcome? (e.g. are blobs downloaded from the remote-execution storage uploaded to the cache storage?)

juergbi replied:
I've added these configuration tests.

Upload to cache storage-service was missing in case two different CAS remotes are configured. I've added this, reusing code we already had for remote execution.

abderrahim reviewed (approved):
Looks good. I have a few comments and questions, but nothing that I would consider a blocker.

juergbi committed:
Drop obsolete cleanup methods

Without direct gRPC connections to remote servers, these cleanup methods
are no longer needed.

juergbi committed:
_protos: Update local_cas.proto from buildbox

juergbi committed:
sandbox: Add helper method to fetch action result outputs

abderrahim commented on src/buildstream/sandbox/_sandboxbuildboxrun.py:
```diff
@@ -148,7 +173,14 @@ def _execute_action(self, action, flags):
-            return remote_execution_pb2.ActionResult().FromString(result_file.read())
+            action_result = remote_execution_pb2.ActionResult().FromString(result_file.read())
+
+        if self.re_remote and context.remote_execution_specs.storage_spec and context.remote_cache_spec:
+            # This ensures that the outputs are uploaded to the cache storage-service
+            # in case different CAS remotes have been configured in the `cache` and `remote-execution` sections.
```
I'm having a hard time wrapping my head around this one. Since we're building locally, shouldn't the outputs be uploaded to the cache storage-service by default? Then we only need to upload them to the remote-execution storage if we're also uploading the action?

Or is it due to the fact that buildbox-run can only take a single instance for both the nested REAPI and where it stores its output? So it will upload its output to the remote-execution CAS as well and we need to transfer it to the cache CAS?

I agree this is an exotic configuration, and shouldn't block landing this MR. I just wanted to raise it just in case.

juergbi replied:
> Or is it due to the fact that buildbox-run can only take a single instance for both the nested REAPI and where it stores its output? So it will upload its output to the remote-execution CAS as well and we need to transfer it to the cache CAS?

That's exactly the reason. While latest buildbox-run actually supports a separate instance for nested REAPI, this is not currently configurable in BuildStream (but it probably makes sense to add this configuration in a follow-up).
